### ISSUE TITLE
False positive with a constrained generic type in a union

### ISSUE THREAD

**Issue by russelldavis** [microsoft/pyright#875]
```python
from typing import Any, Optional, TypeVar
Constrained = TypeVar("Constrained", str, bytes)
def foo1(arg1: Constrained, arg2: Constrained): ...
def foo2(arg1: Optional[Constrained], arg2: Optional[Constrained]): ...

def repro() -> None:
    any_: Any = ""
    foo1("", any_) # No error
    foo1(any_, "") # No error
    foo2("", any_) # No error
    foo2(any_, "") # Error (false positive)
```
Here's the reported error:

> Argument of type "Literal[\'\']" cannot be assigned to parameter "arg2" of type "Constrained | None" in function "foo2"
  Type "str" is not compatible with constrained type "Any"
  Cannot assign to "None"

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in Pyright 1.1.56, which I just published. It will also be in the next version of Pylance.
